### ISSUE TITLE
Consider reducing IDs from 32 bytes to 16

### ISSUE THREAD

**Issue by cjpatton** [ietf-wg-ppm/draft-ietf-ppm-dap#349]
The draft defines the following types as `uint8_t[32]`:
* TaskID
* BatchID
* AggregationJobID

I believe the draft RECOMMENDS that most of these be generated randomly. However, the spec permits different choices here, as the main requirement is *uniqueness*, not necessarily randomness/unpredictability. For example, it is allowed (and perfectly reasonable) to set TaskID to the hash of some unique "context" agreed upon by all of the parties.

I think 32 bytes is overkill for these purposes.

Why 32 bytes? Well, the idea is that an ID could be set to the SHA-256 hash of some string and rely on the collision resistance (CR) of SHA-256 for uniqueness. If we had to truncate the hash, say to 16 bytes, then we could no longer rely on CR for SHA-256, since finding a collision against truncated SHA-256 might be significantly easier.

What's the impact of an ID collision?:
* TaskID: An encrypted input share intended for one task might get decrypted and consumed in another. But if the VDAF is robust, then in the input share will get rejected with high probability. On the other hand, for DAFs (VDAF sans verification), aggregate results would get garbled.
* BatchID: It's not clear which aggregate result to return the Collector.
* AggregationJobID: Helper ends up iin a weird state, causing an aggregation job to fail.

In any case, it is still the responsibility of the parties in the protocol to make sure that all IDs are unique in the relevant context.  In particular, the impacts above are ruled out by a correct implementation.

Assuming you buy this argument, the question that remains is: Why should we bother shortening this? We don't save much in terms of communication overhead. My primary consideration here is implementation: A 16-byte ID is much easier to stuff into a log than a 32-byte ID.

**Comment by wangshan** [comment]
@cjpatton A related issue: report ID is defined differently than the others:
```
opaque TaskID[32];
opaque BatchID[32];
uint8 ReportID[16];
opaque AggregationJobID[32];
```
should it be a uint8 type or opaque as well

**Comment by cjpatton** [comment]
Yeah, that was intentional: ReportID is the new name for the 16-byte random nonce we had in the report metadata. We definitely don't need 32 bytes for this!

**Comment by cjpatton** [comment]
I'm going to suggest that we shorten BatchID and AggregationJobID to 16 bytes, but leave TaskID at 32 bytes in case anyone has concerns about this. I'll send a PR today.
